core/numeric.rs

```diff
@@ -0,0 +1,564 @@
+            (_, NullableInteger::Integer(0)) => NullableInteger::Null,
+            (lhs, NullableInteger::Integer(-1)) => lhs % NullableInteger::Integer(1),
+            (NullableInteger::Integer(lhs), NullableInteger::Integer(rhs)) => {
+                NullableInteger::Integer(lhs % rhs)
+            }
+        }
+    }
+}
+
+// Maximum u64 that can survive a f64 round trip
+const MAX_EXACT: u64 = u64::MAX << 11;
+
+const VERTICAL_TAB: char = '\u{b}';
+
+#[derive(Debug, Clone, Copy)]
+struct DoubleDouble(f64, f64);
+
+impl From<u64> for DoubleDouble {
+    fn from(value: u64) -> Self {
+        let r = value as f64;
+
+        // If the value is smaller than MAX_EXACT, the error isn't significant
+        let rr = if r <= MAX_EXACT as f64 {
+            let round_tripped = value as f64 as u64;
+            let sign = if value >= round_tripped { 1.0 } else { -1.0 };
+
+            // Error term is the signed distance of the round tripped value and itself
+            sign * value.abs_diff(round_tripped) as f64
+        } else {
+            0.0
+        };
+
+        DoubleDouble(r, rr)
+    }
+}
+
+impl From<DoubleDouble> for f64 {
+    fn from(DoubleDouble(a, aa): DoubleDouble) -> Self {
+        a + aa
+    }
+}
+
+impl std::ops::Mul for DoubleDouble {
+    type Output = Self;
+
+    /// Double-Double multiplication.  (self.0, self.1) *= (rhs.0, rhs.1)
+    ///
+    /// Reference:
+    ///   T. J. Dekker, "A Floating-Point Technique for Extending the Available Precision".
+    ///   1971-07-26.
+    ///
+    fn mul(self, rhs: Self) -> Self::Output {
+        // TODO: Better variable naming
+
+        let mask = u64::MAX << 26;
+
+        let hx = f64::from_bits(self.0.to_bits() & mask);
+        let tx = self.0 - hx;
+
+        let hy = f64::from_bits(rhs.0.to_bits() & mask);
+        let ty = rhs.0 - hy;
+
+        let p = hx * hy;
+        let q = hx * ty + tx * hy;
+
+        let c = p + q;
+        let cc = p - c + q + tx * ty;
+        let cc = self.0 * rhs.1 + self.1 * rhs.0 + cc;
+
+        let r = c + cc;
+        let rr = (c - r) + cc;
+
+        DoubleDouble(r, rr)
+    }
+}
+
+impl std::ops::MulAssign for DoubleDouble {
+    fn mul_assign(&mut self, rhs: Self) {
+        *self = self.clone() * rhs;
+    }
+}
+
+pub fn atoi(input: impl AsRef<str>) -> i64 {
+    let input = input
+        .as_ref()
+        .trim_matches(|ch: char| ch.is_ascii_whitespace() || ch == VERTICAL_TAB);
+
+    let mut iter = input.chars().enumerate().peekable();
+
+    iter.next_if(|(_, ch)| matches!(ch, '+' | '-'));
+    let Some((end, _)) = iter.take_while(|(_, ch)| ch.is_ascii_digit()).last() else {
+        return 0;
+    };
+
+    input[0..=end]
+        .parse::<i64>()
+        .unwrap_or_else(|err| match err.kind() {
+            std::num::IntErrorKind::PosOverflow => i64::MAX,
+            std::num::IntErrorKind::NegOverflow => i64::MIN,
+            _ => 0,
+        })
+}
+
+pub fn atof(input: impl AsRef<str>) -> Option<(NonNan, bool)> {
```

Comment:
nit: can we just call this `str_to_float()`? also, i'd prefer this returns a struct

```rust
StrToFloat {
  value: NonNan,
  is_fractional: bool
}
```